docs/design.md

```diff
@@ -0,0 +1,780 @@
+much as possible. This can be done by both limiting the number of opcodes in the
+instruction set, and by factorizing as much as possible several opcodes into the
+same component.
+
+The required degree of the constraints can also influence the number of columns.
+Actually, the maximum degree of the constraints influences the size of the
+evaluation domain, and adding constraints with a higher degree will double its
+size. Hence it is always better to just add intermediate columns to reduce the
+degree of the constraints.
+
+Overall, the goal is to use as few columns as possible, and to keep the degree
+of the constraints as low as possible, which can in turn require more columns.
+
+### Minimal instruction set \label{sec:minimal-instruction-set} {#minimal-instruction-set}
+
+We now present in this section a minimal instruction set.
+
+**Control Flow**:
+
+- `CallAbsImm`, `Ret`: Function call/return management
+- `JmpRelImm`, `JnzFpImm`: Intra-frame jumps
+
+**Arithmetic Operations**:
+
+- `StoreAdd`, `StoreSub`, `StoreMul`, `StoreDiv`: Arithmetic with memory storage
+
+**Memory Operations**:
+
+- `StoreImm`: Store immediate (bytecode) value into memory
+- `StoreDoubleDeref`: Store dereferenced value
+- `StoreToDoubleDeref`: Store at dereferenced address
+
+This proposed instruction set fits in a total of XXX columns. See the
+[Opcodes columns](#opcodes-columns) section for more details.
+
+### Extensions \label{sec:extensions} {#extensions}
+
+If the proposed instruction set is enough to perform any kind of computation,
+one may want to extend it with more opcodes. The purpose of extensions is to
+make some complex operations native to the prover, i.e. to give them directly a
+circuit representation. Whether extensions actually make the whole proving steps
+faster depends on the context and the actual optimization they allow.
+
+Among the most common extensions, we describe below the case of adding different
+"native" types to the instruction set and built-in hash functions.
+
+#### Uint Types \label{sec:uint-types} {#uint-types}
+
+At the prover level, the only native type is the field element. However, at the
+software level, the most common native types are `u32` or `u64`. While it is
+possible
+[to emulate for example a `u256` at the software level](https://github.com/starkware-libs/cairo-lang/blob/v0.14.0.1/src/starkware/cairo/common/uint256.cairo),
+it may be more efficient to instead manage it at the AIR level. For example,
+creating a `u32` with a struct holding two field elements would require two
+memory accesses per variable use instead of one.
+
+At the software level, the main difference between a uint of a given size and a
+felt lies mainly in the division operation. In fact, at least in release mode,
+uints silently overflow and wrap around, behaving like a field element over
+$2^n$. On the other hand, the division for field elements is always exact (every
```

Comment:
I find it weird to talk about "release mode" here out of the blue, which assumes that "At the software level" implies writing Rust